design-documents/graph-ql/path-validator.md

```diff
@@ -0,0 +1,100 @@
+# Magento Framework validators
+The purpose of this document is to a validator for relative url routes not only full urls commonly referred as absolute url routes.
+
+## Terminology
+`UrlValidator` - The existing framework validator `\Magento\Framework\Url\Validator`
+
+`RouteValidator` - The proposed `Magento\Framework\Url\RouteValidator`
+
+## Overview
+### Current situation
+In the Magento 2 framework we can only validate URLs `http://www.somedomain.com/some/path`.
+We don't have the possibility to configure this validator to test relative urls, paths as Zend framework refers to them.
+We should have such kind of validator available for cases where we would just need to validate a relative path and also
+eliminate the directory traversal when such an input comes from the user. 
+
+### Problem
+Such situation of validating a url route did occur with the proposal
+[Domain Whitelist for Configurable 3rd Party Redirects](https://github.com/magento/architecture/pull/204).
+We just don't have a service class to validate url routes, for code re-usage, with that can be injected in all resolvers or
+classes that need this validator such as PayPal Express and others.
+
+### Proposed solution
+As this validator doesn't have anything to do with GraphQl, the best place for it is to add a new validator in framework
+similar to existing `\Magento\Framework\Url\Validator`. So this new class in framework will need approval.
+
+Looking at the implementation of the `\Magento\Framework\Url\Validator` it's based on `\Zend_Validate_Abstract` so it
+complies with our standard of not using native php functions such as parse `parse_url`. Nor we would want to use regex.
+
+
+The new class `RouteValidator` will look like the following and it's namespace will be `Magento\Framework\Url`:
+```   
+   namespace Magento\Framework\Url;
+   
+   /**
+    * Validate url routes
+    */
+   class RouteValidator extends \Zend_Validate_Abstract
```

Comment:
It introduces an additional dependency to Zend Framework 1, which we are trying to avoid.

Reply:
I propose to introduce RouteValidatorInterface, so in the future, we will be able to eliminate inheritance from \Zend_Validate_Abstract without the necessity to support its public methods.

Reply:
We had \Magento\Framework\Validator\ValidatorInterface, just need to implement it

Reply:
yes, it seems that we need a new class that can be configured to be used in both url and path cases
Note that a path is not really any string, and there's some validations involved for delimiters that make sense for a path:
https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Examples_2
https://stackoverflow.com/questions/2163803/what-is-the-semicolon-reserved-for-in-urls